File: libs/ruby/test/test_metric_calendar.rb
require 'minitest/autorun'
require 'date'
require_relative '../lib/metric_calendar'

class TestMetricCalendar < Minitest::Test
  def test_vigil
    r = MetricCalendar.gregorian_to_metric(Date.new(2026, 3, 20))
    assert_equal 56, r.year
    assert r.is_turning
    assert_equal 'Vigil', r.special_day
  end

  def test_balance
    r = MetricCalendar.gregorian_to_metric(Date.new(2026, 3, 21))
    assert r.is_turning
    assert_equal 'Balance', r.special_day
  end

  def test_dawn
    r = MetricCalendar.gregorian_to_metric(Date.new(2026, 3, 22))
    assert r.is_turning
    assert_equal 'Dawn', r.special_day
  end

  def test_unil_1
    r = MetricCalendar.gregorian_to_metric(Date.new(2026, 3, 23))
    assert_equal 56, r.year
    assert_equal 1, r.month
    assert_equal 'Unil', r.month_name
    assert_equal 1, r.day
    assert_equal 1, r.week_day
    assert_equal 'Primday', r.day_name
    refute r.is_rest
    assert_equal 1, r.week
  end

  def test_unil_10
    r = MetricCalendar.gregorian_to_metric(Date.new(2026, 4, 1))
    assert_equal 1, r.month
    assert_equal 10, r.day
    assert_equal 10, r.week_day
    assert_equal 'Decday', r.day_name
    assert r.is_rest
  end

  def test_yule_eve
    r = MetricCalendar.gregorian_to_metric(Date.new(2026, 12, 18))
    assert_equal 56, r.year
    assert r.is_yule
    assert_equal 'Yule Eve', r.special_day
  end

  def test_pre_equinox
    r = MetricCalendar.gregorian_to_metric(Date.new(2025, 1, 1))
    assert_equal 54, r.year  # before March 2025 equinox -> year 54
    assert_equal 10, r.month
    assert_equal 'Decil', r.month_name
    assert_equal 13, r.day
    assert_equal 3, r.week_day
    assert_equal 'Triday', r.day_name
    refute r.is_rest
  end

  def test_midsummer
    r = MetricCalendar.gregorian_to_metric(Date.new(2026, 6, 21))
    assert r.is_midsummer
    assert_equal 4, r.month
    assert_equal 1, r.day
  end

  def test_is_rest_day
    assert MetricCalendar.is_rest_day(Date.new(2026, 4, 1))
    refute MetricCalendar.is_rest_day(Date.new(2026, 3, 23))
  end

  def test_metric_to_gregorian_month
    d = MetricCalendar.metric_to_gregorian(56, 'month', 1, 1)
    assert_equal Date.new(2026, 3, 23), d
  end

  def test_metric_to_gregorian_turning
    d = MetricCalendar.metric_to_gregorian(56, 'turning', 0)
    assert_equal Date.new(2026, 3, 20), d
  end

  def test_metric_to_gregorian_yule
    d = MetricCalendar.metric_to_gregorian(56, 'yule', 0)
    assert_equal Date.new(2026, 12, 18), d
  end

  def test_kindling_requires_leap
    assert_raises(ArgumentError) { MetricCalendar.metric_to_gregorian(56, 'yule', 2) }
  end

  def test_season_index
    r = MetricCalendar.gregorian_to_metric(Date.new(2026, 3, 23))  # Month 1
    assert_equal 0, r.season_index
    r2 = MetricCalendar.gregorian_to_metric(Date.new(2026, 6, 21))  # Month 4
    assert_equal 1, r2.season_index
  end
end

File: libs/python/tests/test_metric_calendar.py
import unittest
import datetime
import sys, os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'src'))
from metric_calendar import gregorian_to_metric, metric_to_gregorian, is_rest_day

class TestGregorianToMetric(unittest.TestCase):

    def test_vigil(self):
        r = gregorian_to_metric(datetime.date(2026, 3, 20))
        self.assertEqual(r.year, 56)
        self.assertTrue(r.is_turning)
        self.assertEqual(r.special_day, 'Vigil')

    def test_balance(self):
        r = gregorian_to_metric(datetime.date(2026, 3, 21))
        self.assertTrue(r.is_turning)
        self.assertEqual(r.special_day, 'Balance')

    def test_dawn(self):
        r = gregorian_to_metric(datetime.date(2026, 3, 22))
        self.assertTrue(r.is_turning)
        self.assertEqual(r.special_day, 'Dawn')

    def test_unil_1(self):
        r = gregorian_to_metric(datetime.date(2026, 3, 23))
        self.assertEqual(r.year, 56)
        self.assertEqual(r.month, 1)
        self.assertEqual(r.month_name, 'Unil')
        self.assertEqual(r.day, 1)
        self.assertEqual(r.week_day, 1)
        self.assertEqual(r.day_name, 'Primday')
        self.assertFalse(r.is_rest)
        self.assertEqual(r.week, 1)

    def test_unil_10(self):
        r = gregorian_to_metric(datetime.date(2026, 4, 1))
        self.assertEqual(r.month, 1)
        self.assertEqual(r.day, 10)
        self.assertEqual(r.week_day, 10)
        self.assertEqual(r.day_name, 'Decday')
        self.assertTrue(r.is_rest)

    def test_yule_eve(self):
        r = gregorian_to_metric(datetime.date(2026, 12, 18))
        self.assertEqual(r.year, 56)
        self.assertTrue(r.is_yule)
        self.assertEqual(r.special_day, 'Yule Eve')

    def test_pre_equinox(self):
        # Jan 1 2025 is before the March 20 2025 equinox, so it belongs to
        # Metric year 54 (which runs from March 20 2024 to March 19 2025).
        r = gregorian_to_metric(datetime.date(2025, 1, 1))
        self.assertEqual(r.year, 54)
        self.assertEqual(r.month, 10)
        self.assertEqual(r.month_name, 'Decil')
        self.assertEqual(r.day, 13)
        self.assertEqual(r.week_day, 3)
        self.assertEqual(r.day_name, 'Triday')
        self.assertFalse(r.is_rest)

    def test_midsummer(self):
        r = gregorian_to_metric(datetime.date(2026, 6, 21))
        self.assertTrue(r.is_midsummer)
        self.assertEqual(r.month, 4)
        self.assertEqual(r.day, 1)

    def test_is_rest_day(self):
        self.assertTrue(is_rest_day(datetime.date(2026, 4, 1)))
        self.assertFalse(is_rest_day(datetime.date(2026, 3, 23)))

    def test_metric_to_gregorian_month(self):
        d = metric_to_gregorian(56, 'month', 1, 1)
        self.assertEqual(d, datetime.date(2026, 3, 23))

    def test_metric_to_gregorian_turning(self):
        d = metric_to_gregorian(56, 'turning', 0)
        self.assertEqual(d, datetime.date(2026, 3, 20))

    def test_metric_to_gregorian_yule(self):
        d = metric_to_gregorian(56, 'yule', 0)
        self.assertEqual(d, datetime.date(2026, 12, 18))

    def test_kindling_requires_leap(self):
        with self.assertRaises(ValueError):
            metric_to_gregorian(56, 'yule', 2)  # Year 56 is not a leap year

if __name__ == '__main__':
    unittest.main()
